src/sidebar/components/test/thread-test.js

```diff
@@ -0,0 +1,284 @@
+import { mount } from 'enzyme';
+import { createElement } from 'preact';
+import { act } from 'preact/test-utils';
+
+import Thread from '../thread';
+import { $imports } from '../thread';
+
+import { checkAccessibility } from '../../../test-util/accessibility';
+import mockImportedComponents from '../../../test-util/mock-imported-components';
+
+// Utility functions to build nested threads
+let lastThreadId = 0;
+const createThread = () => {
+  lastThreadId++;
+  return {
+    id: lastThreadId.toString(),
+    annotation: {},
+    children: [],
+    parent: undefined,
+    collapsed: false,
+    visible: true,
+    depth: 0,
+    replyCount: 0,
+  };
+};
+
+const addChildThread = parent => {
+  const childThread = createThread();
+  childThread.parent = parent.id;
+  parent.children.push(childThread);
+  return childThread;
+};
+
+// NB: This logic lifted from `build-thread.js`
+function countRepliesAndDepth(thread, depth) {
+  const children = thread.children.map(function(child) {
+    return countRepliesAndDepth(child, depth + 1);
+  });
+  return Object.assign({}, thread, {
+    children: children,
+    depth: depth,
+    replyCount: children.reduce(function(total, child) {
+      return total + 1 + child.replyCount;
+    }, 0),
+  });
+}
+
+/**
+ * Utility function: construct a thread with several children
+ */
+const buildThreadWithChildren = () => {
+  let thread = createThread();
+  addChildThread(thread);
+  addChildThread(thread);
+  addChildThread(thread.children[0]);
+  addChildThread(thread.children[0].children[0]);
+  addChildThread(thread.children[1]);
+  // `depth` and `replyCount` are computed properties...
+  thread = countRepliesAndDepth(thread, 0);
+  return thread;
+};
+
+describe('Thread', () => {
+  let fakeStore;
+  let fakeThreadsService;
+  let fakeThreadUtil;
+
+  // Because this is a recursive component, for most tests, we'll want single,
+  // flat `thread` object (so we are not misled by rendered children)
+  const createComponent = props => {
+    return mount(
+      <Thread
+        showDocumentInfo={false}
+        thread={createThread()}
+        threadsService={fakeThreadsService}
+        {...props}
+      />
+    );
+  };
+
+  beforeEach(() => {
+    fakeStore = {
+      setCollapsed: sinon.stub(),
+    };
+
+    fakeThreadsService = {
+      forceVisible: sinon.stub(),
+    };
+
+    fakeThreadUtil = {
+      countHidden: sinon.stub(),
+      countVisible: sinon.stub(),
+    };
+
+    $imports.$mock(mockImportedComponents());
+    $imports.$mock({
+      '../store/use-store': callback => callback(fakeStore),
+      '../util/thread': fakeThreadUtil,
+    });
+  });
+
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  context('thread not at top level (depth > 0)', () => {
```

Comment:
In writing these tests, I realized that there are at least three different ways that client code determines that a given annotation or thread is a "reply":

* `util.annotation_metadata.isReply(annotation)`
* `!!thread.parent`
* `thread.depth > 0`

These all have subtly different meanings.